lib/puppet/util/windows/eventlog.rb

```diff
@@ -0,0 +1,180 @@
+require 'ffi'
+
+# Puppet::Util::Windows::EventLog needs to be requirable without having loaded
+# any other parts of Puppet so it can be leveraged independently by the code
+# that runs Puppet as a service on Windows.
+#
+# For this reason we:
+# - Define Puppet::Util::Windows
+# - Replicate logic that exists elsewhere in puppet/util/windows
+# - Raise generic RuntimeError instead of Puppet::Util::Windows::Error if its not defined
+module Puppet; module Util; module Windows ; end ; end ; end
+
+class Puppet::Util::Windows::EventLog
+  extend FFI::Library
+
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa363679(v=vs.85).aspx
+  EVENTLOG_ERROR_TYPE       = 0x0001
+  EVENTLOG_WARNING_TYPE     = 0x0002
+  EVENTLOG_INFORMATION_TYPE = 0x0004
+
+  # Register an event log handle for the application
+  # @param source_name [String] the name of the event source to retrieve a handle for
+  # @return [void]
+  # @api public
+  def initialize(source_name = 'Puppet')
+    @eventlog_handle = RegisterEventSourceW(nil, wide_string(source_name))
+    if @eventlog_handle == NULL_HANDLE
+      raise EventlogError.new("failed to open Windows eventlog")
+    end
+  end
+
+  # Close this instance's event log handle
+  # @return [void]
+  # @api public
+  def close
+    result_of_close = DeregisterEventSource(@eventlog_handle)
+    if result_of_close == WIN32_FALSE
+      raise EventlogError.new("failed to close Windows eventlog")
+    end
+    @eventlog_handle = nil
+  end
+
+  # Report an event to this instance's event log handle. Accepts a string to
+  #   report (:data => <string>) and event type (:event_type => FixNum) and id
+  # (:event_id => FixNum) as returned by #to_native. The additional arguments to
+  # ReportEventW seen in this method aren't exposed - though ReportEventW
+  # technically can accept multiple strings as well as raw binary data to log,
+  # we accept a single string from Puppet::Util::Log
+  #
+  # @param args [Hash{Symbol=>Object}] options to the associated log event
+  # @return [void]
+  # @api public
+  def report_event(args = {})
+    raise ArgumentError, "data must be a string, not #{args[:data].class}" unless args[:data].is_a?(String)
+    from_string_to_wide_string(args[:data]) do |message_ptr|
+      FFI::MemoryPointer.new(:pointer, 2) do |message|
+        message[0].write_pointer(message_ptr)
+        user_sid, raw_data = nil
+        raw_data_size = 0
+        num_strings = 1
+        eventlog_category = 0
+        report_result = ReportEventW(@eventlog_handle, args[:event_type],
+          eventlog_category, args[:event_id], user_sid,
+          num_strings, raw_data_size, message, raw_data)
+
+        if report_result == WIN32_FALSE
+          raise EventlogError.new("failed to report event to Windows eventlog")
```

Comment:
I don't believe so. Instantiating and raising a Puppet::Util::Windows::Error by itself doesn't trigger an event log write. A quick test confirmed no infinite loop.

```diff
diff --git a/lib/puppet/util/windows/eventlog.rb b/lib/puppet/util/windows/eventlog.rb
index ead6740..29e52af 100644
--- a/lib/puppet/util/windows/eventlog.rb
+++ b/lib/puppet/util/windows/eventlog.rb
@@ -59,9 +59,9 @@ class Puppet::Util::Windows::EventLog
         raw_data_size = 0
         num_strings = 1
         eventlog_category = 0
-        report_result = ReportEventW(@eventlog_handle, args[:event_type],
-          eventlog_category, args[:event_id], user_sid,
-          num_strings, raw_data_size, message, raw_data)
+        report_result = WIN32_FALSE # ReportEventW(@eventlog_handle, args[:event_type],
+          # eventlog_category, args[:event_id], user_sid,
+          # num_strings, raw_data_size, message, raw_data)

         if report_result == WIN32_FALSE
           raise EventlogError.new("failed to report event to Windows eventlog")
```

```ruby
[1] pry(main)> development\puppet(PUP-5756/master/remove_dependency_on_eventlog)> be pry
[2] pry(main)> require 'puppet'
=> true
[3] pry(main)> Puppet::Util::Log.setup_default
=> #<Puppet::Util::Log::DestEventlog:0x00000006748190 @eventlog=#<Puppet::Util::Windows::EventLog:0x00000006748118 @eventlog_handle=151388168>>
[4] pry(main)> Puppet.warning("foo")
Puppet::Util::Windows::Error: failed to report event to Windows eventlog:  The operation completed successfully.
from C:/Users/moses/development/puppet/lib/puppet/util/windows/eventlog.rb:67:in `block (2 levels) in report_event'
[5] pry(main)>
```

Looking up the call list through Puppet::Util::Log and Puppet::Util::Logging, it doesn't appear a failure to log an event generates a new log entry regarding that failure. A failure to set up the logging destination does (via `Puppet.log_exception`, [here](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/log.rb#L145-L150), which will be triggered if we fail to open the eventlog handle. With no destination, that message will get queued. However that's in a rescue that also ensures the console destination is added via `newdestination` after we `log_exception` which will flush the queue and log that to the console. So I think we're ok.